File: ikv-go-client/document.go
package ikvclient

import (
	"errors"

	schemas "github.com/pushkarmoi/ikv-store/ikv-go-client/schemas"
)

// Represents an indexable document for IKV (collection of fields associated with a primary key).
// See IKVDocumentBuilder to build documents.
type IKVDocument struct {
	fields map[string]*schemas.FieldValue
}

type IKVDocumentBuilder struct {
	fields map[string]*schemas.FieldValue
	err    error
}

// Constructor for a new document builder.
func NewIKVDocumentBuilder() *IKVDocumentBuilder {
	return &IKVDocumentBuilder{fields: make(map[string]*schemas.FieldValue)}
}

// Insert a string field.
func (builder *IKVDocumentBuilder) PutStringField(fieldname string, value string) *IKVDocumentBuilder {
	if fieldname == "" || value == "" {
		builder.err = errors.New("missing field name or field value")
		return nil
	}

	builder.fields[fieldname] = &schemas.FieldValue{FieldType: schemas.FieldType_STRING, Value: []byte(value)}
	return builder
}

// Insert a bytes field.
func (builder *IKVDocumentBuilder) PutBytesField(fieldname string, value []byte) *IKVDocumentBuilder {
	if fieldname == "" || len(value) == 0 {
		builder.err = errors.New("missing field name or field value")
		return nil
	}

	builder.fields[fieldname] = &schemas.FieldValue{FieldType: schemas.FieldType_BYTES, Value: value}
	return builder
}

// Build this document.
func (builder *IKVDocumentBuilder) Build() (IKVDocument, error) {
	return IKVDocument{fields: builder.fields}, builder.err
}

File: ikv-python-client/src/ikv_py/document.py
from typing import Dict
from schemas.common_pb2 import FieldValue, IKVDocumentOnWire
from schemas.common_pb2 import STRING as FieldValueString
from schemas.common_pb2 import BYTES as FieldValueBytes
import utils as ikvutils

class IKVDocument:
    def __init__(self, fields: Dict[str, FieldValue]):
        self.fields: Dict[str, FieldValue] = fields
    
    def len(self):
        if self.fields is None:
            return 0
        return len(self.fields)
    
    def as_ikv_document_on_wire(self) -> IKVDocumentOnWire:
        return IKVDocumentOnWire(document=self.fields)

class IKVDocumentBuilder:
    def __init__(self):
        self.fields: Dict[str, FieldValue] = {}
    
    def put_string_field(self, name: str, value: str) -> 'IKVDocumentBuilder':
        name: str = ikvutils.is_valid_str_or_raise(name)
        value: bytes = ikvutils.is_valid_str_or_raise(value).encode('utf-8')

        field_value: FieldValue = FieldValue(fieldType=FieldValueString, value=value)
        self.fields[name] = field_value

        return self
    
    def put_bytes_field(self, name: str, value: bytes) -> 'IKVDocumentBuilder':
        name: str = ikvutils.is_valid_str_or_raise(name)
        value: bytes = ikvutils.is_valid_bytes_or_raise(value)

        field_value: FieldValue = FieldValue(fieldType=FieldValueBytes, value=value)
        self.fields[name] = field_value
        
        return self

    def build(self) -> IKVDocument:
        return IKVDocument(fields=self.fields)